### PR TITLE
Add `clip` option

### LINKED ISSUE #20: Add `clip` option
Implementation of puppeteer [`clip`](https://github.com/GoogleChrome/puppeteer/blob/v1.13.0/docs/api.md#pagescreenshotoptions) functionality.  
Example:
```
{
    clip: {
        x: 10,
        y: 30,
        height: 300,
        width: 200
    }
}
```

- `x` - `<number>` - x-coordinate of top-left corner of clip area;
- `y` - `<number>` - y-coordinate of top-left corner of clip area;
- `width` - `<number>` - width of clipping area;
- `height` - `<number>` - height of clipping area.

### PR DESCRIPTION
**Pull request by nikkki**
- added `clip` option, which is core puppeteer [feature ](https://github.com/GoogleChrome/puppeteer/blob/v1.13.0/docs/api.md#pagescreenshotoptions);  
- added validation to exclude using `clip` and `element` options together;   
- added description in readme.md for new option;
- added typings;  
- added test;

Fixes #20

### CODE REVIEW

**Comment by nikkki** [comment]
I`ll add new option to [capture-website-cli](https://github.com/sindresorhus/capture-website-cli) ASAP

**Comment by nikkki** [comment]
PR for [capture-website-cli](https://github.com/sindresorhus/capture-website-cli/pull/7) also created

**Comment by AdriaRios** [comment]
Hi guys!
Any ETA?

**Comment by nikkki** [comment]
@sindresorhus conflicts are resolved

File: index.d.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /// <reference lib="dom"/>
 /// <reference types="puppeteer"/>
-import {SetCookie, LaunchOptions, Page, Browser} from 'puppeteer';
+import { SetCookie, LaunchOptions, Page, Browser } from 'puppeteer';
```
Comment (sindresorhus) [inline index.d.ts]: Don't do unrelated changes

File: index.d.ts [hunk 2]
```diff
@@ -94,6 +94,11 @@ declare namespace captureWebsite {
 		*/
 		waitForElement?: string;
 
+		/**
+		An object which specifies clipping region of the page.
```
Comment (sindresorhus) [inline index.d.ts]: No point in mentioning the type. That's already clear from the actual type defs. Instead, elaborate on what a clipping region is.

File: index.js [hunk 3]
```diff
@@ -202,6 +214,15 @@ const captureWebsite = async (url, options) => {
 		screenshotOptions.fullPage = false;
 	}
 
+	if (options.clip) {
+		screenshotOptions.clip = {
+			x: options.clip.x || 0,
+			y: options.clip.y || 0,
+			width: options.clip.width || 100,
+			height: options.clip.height || 100
```
Comment (sindresorhus) [inline index.js]: It doesn't make sense to have defaults for width and height.

File: index.d.ts [hunk 4]
```diff
@@ -245,3 +250,29 @@ declare const captureWebsite: {
 };
 
 export = captureWebsite;
+interface ClipOptions {
```
Comment (sindresorhus) [inline index.d.ts]: This should be inside `declare namespace captureWebsite`. And should be named:
```suggestion
interface BoundingBox {
```

Like in Puppeteer.

File: index.d.ts [hunk 5]
```diff
@@ -245,3 +250,29 @@ declare const captureWebsite: {
...
+	y?: number;
+
+	/**
+	width of clipping area
+	@default 100
+	*/
+	width?: number;
+
+	/**
+	height of clipping area
```
Comment (sindresorhus) [inline index.d.ts]: Please format this and the other like the other existing descriptions.

File: index.js [hunk 6]
```diff
@@ -27,6 +27,16 @@ const getBoundingClientRect = element => {
...
 };
 
+function assert(value, message) {
+	if (!value) {
+		throw new Error(message);
+	}
+}
+
+const validateOptions = options => {
+	assert(!options.clip || !options.element, 'options.clip and options.element are exclusive');
```
Comment (sindresorhus) [inline index.js]:
```suggestion
	assert(!(options.clip && options.element), 'The `clip` and `element` option are mutually exclusive');
```

File: index.d.ts [hunk 7]
```diff
@@ -245,3 +250,29 @@ declare const captureWebsite: {
 };
 
 export = captureWebsite;
+interface ClipOptions {
+	/**
+	x-coordinate of top-left corner of clip area
```
Comment (sindresorhus) [inline index.d.ts]: There should be an empty line after the description.

**Comment by sindresorhus** [comment]
@nikkki Ping :)

**Comment by sindresorhus** [comment]
Closing for lack of response.
